target/linux/realtek/files-5.10/drivers/clk/realtek/clk-rtl83xx.c

```diff
@@ -0,0 +1,461 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/* Realtek RTL83XX clock driver
+ *
+ * Copyright (C) 2022 Markus Stockhausen
+ *
+ * This driver provides basic clock support for RTL838X and RTL8389X SoC.
+ * Currently CPU, memory and LXB clock information can be accessed. 
+ * To make use of the driver add the following clocks section to the DTS.
+ *
+ * #include <dt-bindings/clock/rtl83xx-clk.h>
+ * clocks {
+ *   osc: oscillator {
+ *     compatible = "fixed-clock";
+ *     #clock-cells = <0>;
+ *     clock-frequency = <25000000>;
+ *   };
+ *   ccu: clock-controller {
+ *     compatible = "realtek,rtl838x-clock";
+ *     #clock-cells = <1>;
+ *     clocks = <&osc>;
+ *     clock-names = "ref_clk";
+ *   };
+ * };
+ *
+ * Take care to set the compatible value to match your device. E.g.
+ * either realtek,rtl838x-clock or realtek,rtl839x-clock. Afterwards
+ * one can access/identify the clocks in the DTS with <&ccu CLK_CPU>,
+ * <&ccu CLK_MEM> or <&ccu CLK_LXB>.
+ */
+
+#include <linux/clk-provider.h>
+#include <linux/clkdev.h>
+#include <linux/clk.h>
+#include <linux/cpu.h>
+#include <linux/io.h>
+#include <linux/ioport.h>
+#include <linux/of_address.h>
+#include <linux/slab.h>
+
+#include <dt-bindings/clock/rtl83xx-clk.h>
+#include <asm/mach-rtl838x/mach-rtl83xx.h>
+
+/*
+ * Hardware specific definitions
+ */
+
+static const int rtl838x_pll_base[] = {
+	RTL838X_PLL_CPU_CTRL0,
+	RTL838X_PLL_MEM_CTRL0,
+	RTL838X_PLL_LXB_CTRL0
+};
+
+static const int rtl839x_pll_base[] = {
+	RTL839X_PLL_CPU_CTRL0,
+	RTL839X_PLL_MEM_CTRL0,
+	RTL839X_PLL_LXB_CTRL0
+};
+
+static const int rtl83xx_pll_divn3[] = { 2, 3, 4, 6 };
+static const int rtl83xx_pll_xdiv[] = { 2, 4, 2 };
+
+/*
+ * Clock structures
+ */
+
+int rtl_soc_family;
+
+#define PLL_INFO(_id, _name, _pname)	\
+	{				\
+		.id = _id,		\
+		.name = _name,		\
+		.parent_name = _pname	\
+	}
+
+struct rtl_pll_info {
+	unsigned int id;
+	const char *name;
+	const char *parent_name;
+};
+
+struct rtl_pll {
+	struct clk_hw hw;
+	unsigned int id;
+	spinlock_t lock;
+};
+
+static const struct rtl_pll_info pll_info[] = {
+	PLL_INFO(CLK_CPU, "cpu_clk", "ref_clk"),
+	PLL_INFO(CLK_MEM, "mem_clk", "ref_clk"),
+	PLL_INFO(CLK_LXB, "lxb_clk", "ref_clk")
+};
+
+#define PLL_COUNT ARRAY_SIZE(pll_info)
```

Comment:
since your using this define later, why 'hide' it here; keep it at the top?

Reply:
reordered